### PR TITLE
milvus support run on fpga_2020_09_27_https://github.com/milvus-io/milvus/issues/3773

### LINKED ISSUE #3773: Add support run on FPGA for milvus
**Please state your issue using the following template and, most importantly, in English.**

**Is your feature request related to a problem? Please describe.**
due to the high Parallel capability and large memory in  fpga , I want to run milvus on FPGA , so i suggest to add a feature to support it in milvus, In this model ,currently, we can train index use cpu or gpu , and search on fpga。

**Describe the solution you'd like**
 **solution** :
if we want to support run on fpga, we should add some feature, detail as below,
1 . add a compile option for fpga, we can build a fpga version, such as ./build.sh -a -t Debug
2 .add some fpga config option in config file , such as enable/disable  
3 . implement search algorithm on fpga under this folder(milvus/core/src/index/knowhere/knowhere/index/vector_index/fpga)
4 .implement fpga index cache manage.
5 .implement fpga resource manage
6 .and so on

**Describe alternatives you've considered**
no

**Additional context**
copy Index from cpu to FPGA:
void
FPGAIVFPQ::CopyIndexToFpga(){

    auto fpgaIndex = FpgaIndex::GetInstance();
    auto ivf_index = static_cast<faiss::IndexIVFPQ*>(index_.get());
    ivf_index->make_direct_map();
    fpgaIndex ->setIndex(ivf_index);
    fpgaIndex >CopyIndexToFPGA();  
}
void
FPGAIVFPQ::QueryImpl(int64_t n, const float* data, int64_t k, float* distances, int64_t* labels, const Config& config) {

    std::lock_guard<std::mutex> lk(mutex_);
   try{ 

        auto params = GenParams(config);
         int nprobe =params->nprobe;
        LOG_ENGINE_DEBUG_ <<"nprobe:"<<nprobe<<"n:"<<n<<"k:"<<k;
        auto fpgaIndex = fpga::FpgaInst::GetInstance();
        vector<faiss::Index::idx_t> vlabels(n*k);
        vector<float> vdistances(n*k);
        auto t0 = Elapsed();
        fpgaIndex ->BatchAnnQuery(nprobe,n,(float*)data, k, vlabels, vdistances);
        auto t1= Elapsed();
        LOG_ENGINE_DEBUG_ <<" vlabels size:"<<vlabels.size()<<"search time:"<<t1-t0; 
        auto elems = n * k;
        size_t p_id_size = sizeof(int64_t) * elems;
        size_t p_dist_size = sizeof(float) * elems;      
        memcpy(distances,vdistances.data(),p_dist_size);
        memcpy(labels,vlabels.data(),p_id_size);
   }catch(...)
   {

   }
  }

### PR DESCRIPTION
**Pull request by jackhouchina**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

Fixes #3773

**What this PR does / why we need it:**
support run on fpga

**Special notes for your reviewer:**
train on cpu or gpu ,search on fpga

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: core/src/db/engine/ExecutionEngineImpl.cpp [hunk 1]
```diff
@@ -380,9 +403,18 @@ ExecutionEngineImpl::Serialize() {
 
 Status
 ExecutionEngineImpl::Load(bool to_cache) {
-    index_ = std::static_pointer_cast<knowhere::VecIndex>(cache::CpuCacheMgr::GetInstance()->GetIndex(location_));
-    bool already_in_cache = (index_ != nullptr);
+#ifdef MILVUS_FPGA_VERSION
+    auto  cache_index__ = std::static_pointer_cast<knowhere::VecIndex>(cache::FpgaCacheMgr::GetInstance()->GetIndex(location_));
+    bool already_in_caches = (cache_index__ != nullptr);
+    if (already_in_caches) {
+      LOG_ENGINE_DEBUG_ << "jack fpga cache no need to load again" ;
```
Comment (yhmo) [inline core/src/db/engine/ExecutionEngineImpl.cpp:410]: Don't use personal name in log or error message.
Sometimes we can use personal name in code comment such as "TODO: (xxx)", but for log and message it is not allowed.

**Comment by JinHai-CN** [comment]
0.10.0 is already closed. Please submit the PR to 0.10.4.

**Comment by JinHai-CN** [comment]
By the way, DCO signature is necessary.

File: core/src/db/engine/ExecutionEngineImpl.cpp (hunk 1 above)
Comment (jackhouchina) [reply core/src/db/engine/ExecutionEngineImpl.cpp:410]: sorry for forget to delete it , i will modify.

**Comment by jackhouchina** [comment]
> 0.10.0 is already closed. Please submit the PR to 0.10.4.

ok , i will submit the PR to 0.10.4,thanks!

**Comment by jackhouchina** [comment]
> By the way, DCO signature is necessary.

Hi, what's it is DCO signature, i not find it on internet.

**Comment by JinHai-CN** [comment]
> > By the way, DCO signature is necessary.
> 
> Hi, what's it is DCO signature, i not find it on internet.

Please follow: https://github.com/milvus-io/milvus/blob/master/CONTRIBUTING.md#developer-certificate-of-origin-dco

**Comment by shengjun1985** [comment]
There is a new PR for 0.10.4.
Close it.

**Pull request closed without merging**